src/NumberTheory/vinberg.jl

```diff
@@ -0,0 +1,291 @@
+
+@doc raw"""
+    _check_v0(Q::ZZMatrix, v0::ZZMatrix) -> ZZMatrix
+
+Checks if the inner product of the given `v0` is positive or generates such a `v0` if none is given.
+"""
+function _check_v0(Q::ZZMatrix, v0::ZZMatrix) 
+  if iszero(v0)
+    l = nrows(Q)
+    for signal in 1:100000000
+      v0_ = rand(-30:30, l)
+      if !_is_primitive(v0_)
+        continue
+      end
+      v0 = matrix(ZZ, 1, l, v0_)
+      if (v0*Q*transpose(v0))[1, 1] > 0
+        return v0
+      end
+    end 
+    error("Choose another Q")
+  else
+    @req (v0*Q*transpose(v0))[1, 1] > 0 "v0 has non positive inner product"
+    @req isone(reduce(gcd, v)) "v0 is not primitive"
+    return v0
+  end
+end
+
+@doc raw"""
+    _all_root_lengths(Q::ZZMatrix) -> Vector{ZZRingElem}
+
+If the user does not want any specific root lengths, we take all of them.
+
+For `Q` the matrix representing the hyperbolic reflection lattice and `l` a length to check,
+it is necessary that `l` divides $2.i$ for `l` being a possible root length,  
+with `i` being the level of `Q`, i.e. the last invariant (biggest entry) of the Smith normal form of `Q`.
+Prf: `r` being a root implies that $\frac{2.r.L}{r^2}$ is a subset of $\Z$
+     `r` being primitive implies that $2.i.\Z$ is a subset of $2.r.L$,
+     which is a subset of $(r^2).\Z$, what implies that $r^2$ divides $2.i$
+"""
+function _all_root_lengths(Q::ZZMatrix)
+  l = nrows(Q)
+  S = snf(Q)
+  return (-1) * divisors(2 * S[l, l])
+end
+
+@doc raw"""
+    _check_root_lengths(Q::ZZMatrix, root_lengths::Vector{ZZRingElem}) -> Vector{ZZRingElem}
+
+Check whether the given lengths are possible root lengths.
+Unnecessary roots are sorted out and reported if this is wanted.
+Return only the possible root lengths.
+"""
+function _check_root_lengths(Q::ZZMatrix, root_lengths::Vector{ZZRingElem})
+  l = nrows(Q)
+  S = snf(Q)
+  possible_root_lengths = divisors(2 * S[l, l])
+  if isempty(root_lengths)
+    return possible_root_lengths
+  end
+  result = ZZRingElem[]
+  for r in root_lengths
+    if abs(r) in possible_root_lengths
+      push!(result, r)
+    else
+      @vprintln :Vinberg 1 "$r is not a possible root length"
+    end
+  end
+  @req !isempty(result) "No possible root lengths found"
+  return result
+end
+
+@doc raw"""
+    _distance_indices(upper_bound::ZZRingElem, root_lengths::Vector{ZZRingElem}) -> Vector{Tuple{Int, ZZRingElem}}
+
+Returns all possible tuples $(n, l)$ with `n` a natural number and `l` contained in `root_lengths`,
+sorted by increase of the value $\frac{n^2}{l}$, stopping by `upper_bound`.
+"""
+function _distance_indices(upper_bound::Rational, root_lengths::Vector{ZZRingElem})
+
+  result = Tuple{Int,ZZRingElem}[]
+
+  for l in root_lengths
+    x = isqrt(x0*upper_bound) # consider n^2//l < upper_bound => n < sqrt(l*upper_bound)
+    for n in 1:x
+      push!(result, (n, l))
+    end
+  end
+
+  sort!(result; by=x -> (x[1]^2) // abs(x[2]))
+  return result
+end
+
+@doc raw"""
+    _is_primitive(v)
```

Comment:
The return value of `is_something` methods is usually a bool. 
Therefore a more appropriate name for this method would be something like `_rescale_primitive`.